### PR TITLE
feat(batch-exports): Support nullable on-demand run bounds

### PR DESCRIPTION
**Pull request by tomasfarias**
## Problem

This PR contains all "plumbing" changes required to support data interval placeholders.

In particular: `data_interval_end` must be non-`null` for all `BatchExportRun`s. This is a problem for on-demand hogql-model batch exports: On-demand means no schedule from where to get a `data_interval_end`, and hogql-model means users may omit the `data_interval_end` placeholder completely. So, supporting on-demand hogql-model batch exports requires that we support `null` `data_interval_end`.

## Changes

This required quite a few changes, beyond the obvious migration:
* All inputs shared with on-demand had to be updated to take `str | None`.
  * Notably, non-file-download destination batch exports override this, as they do not support on demand.
  * Not sure if we would ever do, e.g., BigQuery on demand, or it would remain only file-download. So, for now, other destinations can have `data_interval_end: str`.

* Since we don't have an interval, we need a way to derive a timeout for file-download on demand batch exports with hogql-models.
  * The timeout is now passed as an input, hardcoded to 20 minutes for internal stage activity (15 min execution timeout + 5 mins of extra time to handle the error) and to 6 hours for the main activity.
  * This is very arbitrary, and the second one may be too much. My thinking was we can tune both later. The main activity timeout has to be large enough to accommodate potentially any query, maybe we can have it adapt over time later.

* Since we don't have a `data_interval_end`, we now need a different way to name files in S3 for file-download on-demand hogql-model:
  * Before, we identified runs with the interval, now we identify them with their `run_id`. 
  * I added the run's created_at as part of the key, just to have some way of sorting the files, and for the user to be able to distinguish them with something that carries more meaning than a random uuid.

* A bunch of null-checks and assertions everywhere.

## How did you test this code?

There are a bunch of new tests, all are passing.

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Automatic notifications

No

- [ ] Publish to changelog?

## Docs update

No

## 🤖 Agent context

I had a manual draft of this whole stack of PRs, and then prompted astra to add the migration. I thought it would be a lot of mechanical null-checks, but unfortunately it ended up requiring handling a bunch of edge cases like I mentioned above. So, there was a long tail of going back and forth getting things close to what I like.

I also had the robot author tests.

**Autonomy:** Human-driven (agent-assisted) - or - Fully autonomous

**Agent:** tool, model

### CODE REVIEW

File: posthog/settings/batch_exports.py [hunk 1]
```diff
@@ -177,6 +177,9 @@
 BATCH_EXPORT_HOGQL_MAX_EXECUTION_TIME: int = get_from_env(
     "BATCH_EXPORT_HOGQL_MAX_EXECUTION_TIME", 900, type_cast=int
 )  # 15 minutes (deliberately under the file-download stage timeout of 20 minutes)
+BATCH_EXPORT_HOGQL_ON_DEMAND_MAIN_TIMEOUT_SECONDS: int = get_from_env(
+    "BATCH_EXPORT_HOGQL_ON_DEMAND_MAIN_TIMEOUT_SECONDS", 6 * 60 * 60, type_cast=int
+)
```
[the remark this answers is not in the document]
Comment (tomasfarias) [reply posthog/settings/batch_exports.py:182]: This warning is so annoying: There is a default! That's to use! we don't need wiring. And also, this isn't a secret, so why even mention posthog/secrets at all??

[the remark this answers is not in the document]
Comment (tomasfarias) [reply posthog/settings/batch_exports.py:182]: Can't you just use some context or something to figure out when to warn people? I.e. no defaults, actual secrets, empty/none values?

[the remark this answers is not in the document]
Comment (tomasfarias) [reply posthog/settings/batch_exports.py:182]: Yes remember that for as long as you interact with me (which hopefully won't be anywhere else besides here, I despise you).

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py [hunk 2]
```diff
@@ -330,25 +345,30 @@ async def insert_into_internal_stage_activity(
...
-        data_interval_end = dt.datetime.fromisoformat(inputs.data_interval_end)
         full_range = (data_interval_start, data_interval_end)
 
+        if data_interval_end is None and not inputs.on_demand:
+            raise ValueError("Scheduled batch exports require data_interval_end")
+
+        staging_run_id = None
+        if inputs.on_demand and (data_interval_start is None or data_interval_end is None):
+            if inputs.run_id is None:
+                raise ValueError("On-demand batch exports with missing interval bounds require run_id")
```
[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py]: Yeah, I did this somewhat.

**Comment by tomasfarias** [comment]
I don't think data-tools needs to review: the only hogql system table change is a very trivial `null=True` change.

File: products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py [hunk 3]
```diff
@@ -115,8 +115,9 @@ def _is_authorization_failure_response_error(err: HttpResponseError) -> bool:
     return getattr(err, "error_code", None) == StorageErrorCode.AUTHORIZATION_FAILURE
 
 
-@dataclasses.dataclass(kw_only=True)
+@dataclasses.dataclass(frozen=False, kw_only=True)
 class AzureBlobInsertInputs(BatchExportInsertInputs):
+    data_interval_end: str
     container_name: str
```
Comment (tomasfarias) [inline products/batch_exports/backend/temporal/destinations/azure_blob_batch_export.py:109]: `data_interval_end` gets set by each workflow based on the schedule. So we can take advantage of the tighter type signature.

File: products/batch_exports/backend/temporal/destinations/file_download_batch_export.py [hunk 4]
```diff
@@ -248,6 +250,18 @@ async def export_to_file_download_bucket_with_temporary_credentials(inputs: Expo
...
+        inputs.batch_export.data_interval_start is None or inputs.batch_export.data_interval_end is None
+    ):
+        run_id = inputs.batch_export.run_id
+        if run_id is None:
+            raise MissingRequiredInputsError("An on-demand file download requires a run_id")
+        prefix = FILE_DOWNLOAD_UNBOUNDED_PREFIX.format(
+            batch_export_id=inputs.batch_export.batch_export_id, batch_export_run_id=run_id
+        )
+        run = await BatchExportRun.objects.aget(id=run_id, batch_export_on_demand__team_id=inputs.batch_export.team_id)
+        file_name_prefix = f"export-{run.created_at.astimezone(dt.UTC):%Y-%m-%dT%H-%M-%SZ}"
```
Comment (tomasfarias) [inline products/batch_exports/backend/temporal/destinations/file_download_batch_export.py]: I thought having some date in the file name would be useful for, e.g., sorting.

File: products/batch_exports/backend/temporal/pipeline/entrypoint.py [hunk 5]
```diff
@@ -297,8 +321,15 @@ async def execute_batch_export_using_internal_stage(
     heartbeat_timeout_seconds = settings.BATCH_EXPORT_HEARTBEAT_TIMEOUT_SECONDS
     heartbeat_timeout = dt.timedelta(seconds=heartbeat_timeout_seconds) if heartbeat_timeout_seconds else None
 
+    if interval is None:
+        assert model_name == "hogql" and batch_export_inputs.on_demand, (
```
[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: Nothing was replaced this is an entirely new check. IDK what you are talking about. If you are comparing the current diff to back when this was a draft, then yeah, it was a draft I was trying stuff out. Don't do that.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: What? 

> commit https://github.com/PostHog/posthog/commit/df623b718344c8af7e872a7c39c02d39b12b9cd5 (the first commit on this branch) introduced [the change]

> So within the diff I'm reviewing (this PR against master), the check was genuinely replaced, not newly added

These two statements are contradictory. Are you reviewing this PR against itself or this PR against master? 

The change was replaced while the PR was in draft. Why are you even considering a draft state? Compare the PR AS-IS against master and make suggestions if necessary. Ignore any of my drafts.

What an absolute waste of time.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: > enforce that invariant with a bare assert, which Python strips under -O/PYTHONOPTIMIZE

And? Either prove the invariant is violated or GTFO.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: I believe I said "prove it or GTFO". We are now in the part you "GTFO".

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: Resolve the thread then. Clean-up your mess.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: But the finding is resolved, by yourself: You have shown the invariant is not violated, so there is no reason to opt for the more tedious runtime check.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: Oh, so are you able to predict the future? Why not flag that in a future PR when the invariant is actually violated?

Besides: you said yourself you were "dropping it". Not marking this as resolved looks like anything BUT dropping it.

[the remark this answers is not in the document]
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/entrypoint.py:325]: Great, we are in agreement this can be resolved.

File: products/batch_exports/backend/temporal/destinations/file_download_batch_export.py [hunk 6]
```diff
@@ -248,6 +250,18 @@ async def export_to_file_download_bucket_with_temporary_credentials(inputs: Expo
...
+    if inputs.batch_export.on_demand and (
+        inputs.batch_export.data_interval_start is None or inputs.batch_export.data_interval_end is None
+    ):
+        run_id = inputs.batch_export.run_id
+        if run_id is None:
+            raise MissingRequiredInputsError("An on-demand file download requires a run_id")
+        prefix = FILE_DOWNLOAD_UNBOUNDED_PREFIX.format(
+            batch_export_id=inputs.batch_export.batch_export_id, batch_export_run_id=run_id
+        )
+        run = await BatchExportRun.objects.aget(id=run_id, batch_export_on_demand__team_id=inputs.batch_export.team_id)
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/destinations/file_download_batch_export.py]: this database call is run outside of the heartbeater (line 294 below), so I think we just need to hoist that up above

File: products/batch_exports/backend/temporal/destinations/utils.py [hunk 7]
```diff
@@ -58,27 +59,42 @@ def get_absolute_key_prefix(
...
     compression_extension: str | None = None,
     file_number: int = 0,
     include_file_number: bool = False,
+    file_name_prefix: str | None = None,
 ) -> str:
     """Generate object storage key for batch export files."""
-    key_prefix = get_key_prefix(prefix, data_interval_start, data_interval_end, batch_export_model)
-
-    base_file_name = f"{data_interval_start}-{data_interval_end}"
+    if file_name_prefix is not None and (data_interval_start is None or data_interval_end is None):
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/destinations/utils.py:111]: nit: this function feels like it's getting a bit messy

File: products/batch_exports/backend/temporal/destinations/utils.py [hunk 8]
```diff
@@ -58,27 +59,42 @@ def get_absolute_key_prefix(
...
+        key_prefix = prefix
+        base_file_name = file_name_prefix
+        include_file_number = True
+    else:
+        if data_interval_end is None:
+            raise MissingRequiredInputsError("Provide a filename prefix for an export without an end bound.")
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/destinations/utils.py:117]: nit: I think this error ends up visible to the end user and isn't really that informative to them (it's an internal error message)

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py [hunk 9]
```diff
@@ -330,25 +336,30 @@ async def insert_into_internal_stage_activity(
...
-        data_interval_end = dt.datetime.fromisoformat(inputs.data_interval_end)
         full_range = (data_interval_start, data_interval_end)
 
+        if data_interval_end is None and not inputs.on_demand:
+            raise MissingRequiredInputsError("Scheduled batch exports require data_interval_end")
+
+        staging_run_id = None
+        if inputs.on_demand and (data_interval_start is None or data_interval_end is None):
+            if inputs.run_id is None:
+                raise MissingRequiredInputsError("On-demand batch exports with missing interval bounds require run_id")
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/pipeline/internal_stage.py:352]: question: in which cases can `run_id` be missing?

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py [hunk 10]
```diff
@@ -104,9 +108,11 @@ class HogQLQueryResourceLimitExceededError(Exception):
 # run without failing the activity. This mirrors how the destination activities treat their own
 # non-retryable errors (see `handle_non_retryable_errors`), and prevents us being alerted on user errors.
 NON_RETRYABLE_ERRORS: tuple[type[Exception], ...] = (
-    DataIntervalEndInFutureError,
+    DataIntervalInFutureError,
+    MissingRequiredInputsError,
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/pipeline/internal_stage.py]: should this be non-retryable? It feels as though this is a programming error rather than a user error (at least in most of the cases I can see)

File: products/batch_exports/backend/temporal/errors.py [hunk 11]
```diff
@@ -0,0 +1,2 @@
+class MissingRequiredInputsError(Exception):
```
Comment (rossgray) [inline products/batch_exports/backend/temporal/errors.py]: as mentioned elsewhere, it's not clear to me if this is an internal error or a user-facing one. It feels like it's an internal one because we should be catching missing user inputs earlier (eg at the API-level). But would be good to add a docstring to say what this represents

**Review by rossgray — CHANGES_REQUESTED** [review]
Although a bit of a pain, I agree that supporting nullable on-demand run bounds is the right thing to do.

Changes look good, though I had a few questions and comments.

One other more general thought is that we're now doing a lot of defensive type checking everywhere (eg checking bounds are not null here and there). I don't think this is necessarily a problem and maybe it is inevitable, however I wonder if it is a sign that we should introduce some new abstractions at some point (eg concepts of a bounded run, unbounded run).

File: products/batch_exports/backend/temporal/destinations/file_download_batch_export.py (hunk 6 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/destinations/file_download_batch_export.py]: Should be fairly fast since it's just a select by id, but yeah it's better to have it under heartbeater in case the db is being slow for whatever reason.

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py (hunk 9 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py:352]: It can't really. It's only if we ever introduce a bug here.

Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py:352]: We could replace these with assertions. Or make `MissingRequiredInputsError` retryable so we'll see the batch export retry continuously and we can handle the bug.

File: products/batch_exports/backend/temporal/destinations/utils.py (hunk 8 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/destinations/utils.py:117]: Yeah, similar to what I commented below: 

> We could replace these with assertions. Or make MissingRequiredInputsError retryable so we'll see the batch export retry continuously and we can handle the bug.

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py (hunk 10 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py]: Yeah, I'll make it retryable, it's the same issue as the other two comments.

File: products/batch_exports/backend/temporal/errors.py (hunk 11 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/errors.py]: I'll make it retryable. It's there mostly to catch a runtime situation that should be impossible. At the moment, me (and the robots) don't see a path in which this would be raised. It's more just there for the future.

Ultimately, we have accumulated some tech debt here: we should have a properly typed dataclass that has every input we need and nothing else, so there's never anything "missing".

But anyways, for now I'll make this retryable + add some docstring that this is just an internal broken state.

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py (hunk 9 above)
Comment (rossgray) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py:352]: could we make the type non-nullable too? (I assume there was probably a reason we made it optional?)

File: products/batch_exports/backend/temporal/errors.py (hunk 11 above)
Comment (rossgray) [reply products/batch_exports/backend/temporal/errors.py]: yeah, I think that's the most pragmatic solution for now 👍

File: products/batch_exports/backend/temporal/pipeline/internal_stage.py (hunk 9 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/pipeline/internal_stage.py:352]: I think it's for older batch exports

File: products/batch_exports/backend/temporal/destinations/file_download_batch_export.py (hunk 6 above)
Comment (tomasfarias) [reply products/batch_exports/backend/temporal/destinations/file_download_batch_export.py]: I've done just that

Comment (tomasfarias) [reply products/batch_exports/backend/temporal/destinations/file_download_batch_export.py]: I also moved the db call in the other activity from the same workflow, which was also outside of heartbeater.

**Review by rossgray — APPROVED** [review]
thanks for addressing my comments. I think this looks good to me now 🚢

**Comment by gantoine** [comment]
🚦 Merge queue triage: **real test failure**

The queue attempt failed the `Python code quality` static typing check (mypy, step "Check static typing"). It failed the same way on both the first attempt and the follow-up bisection attempt on this same commit, and the bisection runs master plus only this PR, so the type errors are in this branch's code rather than a queue teardown or a flake. Requeueing as-is will not help. Cause reconstructed from CI (the Trunk CLI is not available in this environment).

Next step: fix the mypy errors (run `uv run mypy --cache-fine-grained .` the way CI does), push, wait for green checks, then resubmit the stack to the queue. I won't repeat this until the branch or the queue state moves.



---
_Generated by [Claude Code](https://claude.ai/code)_
